File: addon/Achievements.lua
local _, WS = ...

-- Achievement evaluation and rewards. Definitions live in
-- Data\Achievements.lua (WS.AchievementData); this module runs their tests
-- (once per second during a run, on boss kills, and when a run ends), grants
-- rewards, and raises the WoW-styled toast.
WS.Achievements = {}
local Achievements = WS.Achievements

function Achievements:Check()
    local db = WoWSurvivorsDB
    local run = WS.Game.running and WS.Game.run or nil
    local player = WS.Game.running and WS.Game.player or nil
    for i = 1, #WS.AchievementOrder do
        local id = WS.AchievementOrder[i]
        if not db.achievements[id] then
            local data = WS.AchievementData[id]
            if data and data.test(db, run, player) then
                self:Grant(id)
            end
        end
    end
end

function Achievements:Grant(id)
    local db = WoWSurvivorsDB
    if db.achievements[id] then return end
    local data = WS.AchievementData[id]
    db.achievements[id] = true

    WS.Audio:Play("toast")
    WS.UI:Toast(data.icon, "Achievement Earned!", data.name)

    local reward = data.reward
    if reward then
        if reward.type == "character" and not db.unlocks.characters[reward.id] then
            db.unlocks.characters[reward.id] = true
            local character = WS.Characters[reward.id]
            WS.UI:Toast(character.classIcon, "New Survivor Unlocked!", character.name)
        elseif reward.type == "map" and not db.unlocks.maps[reward.id] then
            db.unlocks.maps[reward.id] = true
            local map = WS.Maps[reward.id]
            WS.UI:Toast(map.icon, "New Battlefield Unlocked!", map.name)
        elseif reward.type == "gold" then
            db.gold = db.gold + reward.amount
            WS.UI:Toast("Interface\\Icons\\INV_Misc_Coin_01", "Bounty Collected!", reward.amount .. " gold")
        end
    end

    -- Keep any open menu views current.
    if WS.UI.menu and WS.UI.menu:IsShown() then
        WS.UI:RefreshMenu()
    end
end

File: src/game/achievements.js
/* Achievement checking. Tests read persisted statistics, so an achievement can
 * complete mid-run or at the moment the run's totals are banked. */
'use strict';
(function (WS) {

  const Achievements = {};

  Achievements.check = function () {
    const db = WS.Save.db;
    let earned = null;
    for (const id of WS.AchievementOrder) {
      if (db.achievements[id]) continue;
      const a = WS.Achievements[id];
      let passed = false;
      try { passed = a.test(db); } catch (e) { passed = false; }
      if (!passed) continue;

      db.achievements[id] = true;
      earned = a;
      if (a.reward) {
        if (a.reward.type === 'gold') WS.Save.addGold(a.reward.amount);
        else if (a.reward.type === 'character') db.unlocks.characters[a.reward.id] = true;
        else if (a.reward.type === 'map') db.unlocks.maps[a.reward.id] = true;
      }
      WS.Game.toast('Achievement: ' + a.name, Achievements.rewardText(a));
      WS.Audio.play('level');
    }
    if (earned) WS.Save.save();
    return earned;
  };

  Achievements.rewardText = function (a) {
    if (!a.reward) return a.description;
    if (a.reward.type === 'gold') return `+${a.reward.amount} gold banked.`;
    if (a.reward.type === 'character') return `${WS.Characters[a.reward.id].name} is now playable.`;
    if (a.reward.type === 'map') return `${WS.Maps[a.reward.id].name} is now open.`;
    return a.description;
  };

  WS.AchievementSystem = Achievements;
  // The data table is WS.Achievements; the system hangs off it for convenience
  // so ported call sites (`WS.Achievements.check()`) still read naturally.
  WS.Achievements.check = Achievements.check;
  WS.Achievements.rewardText = Achievements.rewardText;

})(window.WS);
